clients/application-amqp/src/main/java/org/eclipse/hono/application/client/amqp/AmqpApplicationClientFactory.java

```diff
@@ -13,11 +13,78 @@
 
 package org.eclipse.hono.application.client.amqp;
 
+import java.util.function.Function;
+
 import org.eclipse.hono.application.client.ApplicationClientFactory;
+import org.eclipse.hono.application.client.DownstreamMessage;
+import org.eclipse.hono.application.client.MessageConsumer;
 import org.eclipse.hono.util.Lifecycle;
 
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+
 /**
  * A factory for creating clients for Hono's AMQP-based north bound APIs.
  */
 public interface AmqpApplicationClientFactory extends ApplicationClientFactory<AmqpMessageContext>, Lifecycle {
```

Comment:
Good question. Does it make sense to do this for the AMQP based implementation only? As far as I can see there is no comparable behavior in the Kafka based implementation, is there?

Reply:
Yes, this would be for AMQP only because the Kafka client manages the connections internally. I have with the `Lifecycle` anyway a difference between the AMQP-based and the Kafka-based factory. The Kafka-based factory does not have to be started. For places where I want to be able to use both factories, I would cast the AMQP-based one and then start it. I don't really see any alternative to this anyway. Then I can also set a reconnect handler, can't I?

Reply:
The HonoConnectionImpl also handles reconnections automatically (based on configuration properties). 

> Then I can also set a reconnect handler, can't I?

That would be the whole point of it, I guess. I wonder, though, if the Kafka bases implementation could offer the same thing, i.e. registering handlers for when the connection is lost and for when it is being re-established? Or is this done automatically anyway "under the hood"? If so, then we could do this for the AMQP based implementation as well. We are already doing something like that in the legacy command consumers ....

Reply:
> Or is this done automatically anyway "under the hood"?

Yes, it's completely under the hood (which makes especially sense, as it is not only one connection but connections are dynamically opened to specific brokers in the cluster).